### DISCUSSION TITLE
I want to use VCPKG_DISABLE_COMPILER_TRACKING along with compiler checks

### DISCUSSION THREAD

**Discussion by tsondergaard** [microsoft/vcpkg#37399] [General]
Vcpkg ABI hash will change when the compiler is changed even if it is just a revision update from e.g gcc 11.4.1 to 11.4.2. This can be a big problem if your dependencies takes hours to compile and you allow developers to update their compilers independently. With #17526 we got VCPKG_DISABLE_COMPILER_TRACKING, but that turns compiler tracking completely off. What I would like is to be able to use VCPKG_DISABLE_COMPILER_TRACKING, but still tag the ABI hash with some compiler related information. I believe I can do that by simply putting compiler related information in my triplet. Say for instance I want to define x64-linux-gcc-11.cmake it could be done like this, I think:

```
# copied from vcpkg.git/triplets/community/x64-linux-dynamic.cmake:
...
...

# Added to disable precise compiler tracking
set(VCPKG_DISABLE_COMPILER_TRACKING ON)

# Added to encode compiler "profile" in triplet
set(MY_COMPILER GNU)
set(MY_COMPILER_VERSION 11)
```

This mostly works for me, except MY_COMPILER and MY_COMPILER_VERSION are not actually read by anything, it is just added to the triplet, which if I understand it correctly is enough to ensure that it gets a different ABI hash.

Since it is not read by anything it is entirely possible that a build could use this triplet with gcc 13 and then I get my cache populated with gcc 13 generated code, which is not what I want.

Is there are possibility to add inject triplet related checks somewhere?

The triplet itself is a .cmake file. What context is it executed in? Could the triplet itself inject additional code that adds the compiler check? I imagine I don't want to add code to the triplet itself as any changes to that code would affect the ABI hash, right?

How is the ABI hash contribution derived from the triplet file? Is it just the hash of the raw bytes of the file including whitespace and all?

**Comment by Neumann-A** [comment]
Use a toolchain to hardcode `g++11` etc. as `CMAKE_CXX_COMPILER`

**Comment by tsondergaard** [comment]
@Neumann-A, thanks for the quick answer. What would that look like? My `CMakePresets.json` has this configure preset:

```
        {
            "name": "base",
            "hidden": true,
            "binaryDir": "${sourceDir}/build",
            "generator": "Ninja",
            "cacheVariables": {
                "CMAKE_TOOLCHAIN_FILE": "${sourceDir}/vcpkg/scripts/buildsystems/vcpkg.cmake",
                "VCPKG_OVERLAY_TRIPLETS": "${sourceDir}/third-party/vcpkg-triplets"
            }
        }
```

So the CMAKE_TOOLCHAIN_FILE I specify is the vcpkg.cmake supplied one. Can I specify an additional toolchain file somehow?

**Reply by Neumann-A** [reply]
https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_chainload_toolchain_file

**Reply by tsondergaard** [reply]
Thanks I'll give that a try!

**Comment by Osyotr** [comment]
Sounds like another use-case for https://github.com/microsoft/vcpkg-tool/pull/816
